deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java

```diff
@@ -0,0 +1,1044 @@
+        if (voidConfiguration.getControllerAddress() == null)
+            voidConfiguration.setControllerAddress(System.getenv("SPARK_PUBLIC_DNS"));
+
+        if (voidConfiguration.getControllerAddress() == null)
+            throw new DL4JInvalidConfigException("Can't get Spark Master local address. Please specify it manually using VoidConfiguration.setControllerAddress(String) method");
+
+        // first of all, we're instantiating ParameterServer shard here
+        Transport transport = voidConfiguration.getTransportType() == TransportType.ROUTED ? new RoutedTransport() : voidConfiguration.getTransportType() == TransportType.BROADCAST ? new MulticastTransport() : this.transport;
+
+        if (transport == null)
+            throw new DL4JInvalidConfigException("No Transport implementation was defined for this training session!");
+
+        // it's safe to call init on CompGraph, it has isInit check within
+        network.getNetwork().init();
+
+        // this instance will be SilentWorker - it'll accept and apply messages, but won't contribute to training. And we init it only once
+        if (isFirstRun.compareAndSet(false, true)) {
+            trainingDriver = new SilentTrainingDriver(network.getNetwork().params(), network.getNetwork().getOptimizer().getStepFunction());
+            VoidParameterServer.getInstance().init(voidConfiguration, transport, trainingDriver);
+        }
+    }
+
+    protected void prepareNetworkAndStuff(SparkDl4jMultiLayer network) {
+        // first of all, we're instantiating ParameterServer shard here\
+        if (numWorkers == null)
+            numWorkers = network.getSparkContext().defaultParallelism();
+
+        // set current box as controller, if field is unset
+        if (voidConfiguration.getControllerAddress() == null)
+            voidConfiguration.setControllerAddress(System.getenv("SPARK_PUBLIC_DNS"));
+
+        if (voidConfiguration.getControllerAddress() == null)
+            throw new DL4JInvalidConfigException("Can't get Spark Master local address. Please specify it manually using VoidConfiguration.setControllerAddress(String) method");
+
+        // we're forcing proper defaults
+        log.info("Setting controller address to {}:{}", voidConfiguration.getControllerAddress(), voidConfiguration.getUnicastPort());
+        voidConfiguration.setShardAddresses(voidConfiguration.getControllerAddress());
+        voidConfiguration.setNumberOfShards(1);
+
+        Transport transport = voidConfiguration.getTransportType() == TransportType.ROUTED ? new RoutedTransport() : voidConfiguration.getTransportType() == TransportType.BROADCAST ? new MulticastTransport() : this.transport;
+
+        if (transport == null)
+            throw new DL4JInvalidConfigException("No Transport implementation was defined for this training session!");
+
+        // TODO: Alex, any better ideas here? Plus, if we're not on SGD updater, we might want to eventually dump updater to Master?
```

Comment:
Nah that's what we've already discussed and what i'm doing right now: passing back/averaging  updaters states.